### PR TITLE
kv/RocksDBStore: Add admin command to retrieve stats/performance

### PR DESCRIPTION
**Pull request by aclamk**
Added admin command "dump_rocksdb_stats" that extracts RocksDB perf counters,
compaction and column family statistics. Extracted data is formated using
typical CephFormatter for admin commands.

There are 4 flavors of dump_rocksdb_stats:
- all: prints everything that could be extracted from RocksDB
- debug: prints extensive data on performance, column families, shards,
         but omit entries that are useful for RocksDB developers
- objectstore: prints data that allows to inspect load, latencies and sizes of
               column families. Does not split cfs for shards.
- telemetry: prints rudimentary data that still allows to capture state

It requires rocksdb_perf=true option to be set to give back useful data.
Command "dump_rocksdb_stats" can be applied to both OSD and MON.
If multiple RocksDB instances are in one ceph context, only first created can
be monitored.

Signed-off-by: Adam Kupczyk <akupczyk@redhat.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests
- Teuthology
  - [ ] Completed teuthology run
  - [ ] No teuthology test necessary (e.g., documentation)

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by aclamk** [comment]
*DNM* some errors in data aggregation found.

**Comment by ljflores** [comment]
jenkins retest this please

**Review by ljflores — COMMENTED** [review]
I need to take more time to understand the logic of this patch, but I did a preliminary test to check the admin command, and it appears to work as promised! I tried all flavors, but I'll just post the results of the Telemetry flavor for conciseness, if anyone is curious (this is from a typical vstart cluster btw):

```
ceph daemon osd.0 dump_rocksdb_stats telemetry
{
    "main": {
        "rocksdb.blobdb.num.get": 0,
        "rocksdb.blobdb.num.keys.read": 0,
        "rocksdb.blobdb.num.keys.written": 0,
        "rocksdb.blobdb.num.multiget": 0,
        "rocksdb.blobdb.num.next": 0,
        "rocksdb.blobdb.num.prev": 0,
        "rocksdb.blobdb.num.put": 0,
        "rocksdb.blobdb.num.seek": 0,
        "rocksdb.blobdb.num.write": 0,
        "rocksdb.blobdb.write.blob": 0,
        "rocksdb.block.cache.data.hit": 22,
        "rocksdb.block.cache.data.miss": 3,
        "rocksdb.block.cache.filter.add": 3,
        "rocksdb.block.cache.filter.miss": 3,
        "rocksdb.block.cache.hit": 797,
        "rocksdb.block.cache.miss": 9,
        "rocksdb.bytes.read": 882,
        "rocksdb.bytes.written": 8640992,
        "rocksdb.memtable.hit": 4,
        "rocksdb.memtable.miss": 745
    },
    "histogram": {
        "rocksdb.compaction.times.cpu_micros": {
            "p50": 0,
            "p95": 0,
            "p99": 0,
            "avg": 0,
            "std": 0,
            "min": -9223372036854775808,
            "max": 0,
            "count": 0,
            "sum": 0
        },
        "rocksdb.compaction.times.micros": {
            "p50": 0,
            "p95": 0,
            "p99": 0,
            "avg": 0,
            "std": 0,
            "min": -9223372036854775808,
            "max": 0,
            "count": 0,
            "sum": 0
        },
        "rocksdb.db.get.micros": {
            "p50": 12.937500,
            "p95": 33.848696,
            "p99": 49.634516,
            "avg": 16.126836,
            "std": 12.164919,
            "min": 2,
            "max": 227,
            "count": 749,
            "sum": 12079
        },
        "rocksdb.db.write.micros": {
            "p50": 50.937335,
            "p95": 392.406181,
            "p99": 577.801325,
            "avg": 119.587445,
            "std": 206.181337,
            "min": 7,
            "max": 10932,
            "count": 10498,
            "sum": 1255429
        },
        "rocksdb.sst.read.micros": {
            "p50": 0.600000,
            "p95": 2.100000,
            "p99": 2.820000,
            "avg": 0.944444,
            "std": 0.779759,
            "min": 0,
            "max": 3,
            "count": 18,
            "sum": 17
        },
        "rocksdb.wal.file.sync.micros": {
            "p50": 275.890756,
            "p95": 543.008357,
            "p99": 797.844068,
            "avg": 296.639090,
            "std": 127.523004,
            "min": 139,
            "max": 2133,
            "count": 2768,
            "sum": 821097
        }
    },
    "columns": {
        "all_columns": {
            "sum": {
                "KeyDrop": 0,
                "KeyIn": 0,
                "NumFiles": 3
            },
            "total_slowdown": 0,
            "total_stop": 0
        }
    }
}
```

**Comment by ljflores** [comment]
jenkins retest this please

File: src/kv/RocksDBStore_stats.cc [hunk 1]
```diff
@@ -0,0 +1,810 @@
...
+    }
+    dout(5) << __func__ << " start" << dendl;
+    std::map<std::string, uint32_t> histograms_map;
+    for (auto& t: rocksdb::HistogramsNameMap) {
+      histograms_map.emplace(t.second, t.first);
+    }
+    for (auto& c: histograms) {
+      auto it = histograms_map.find(c.name);
+      if (it != histograms_map.end()) {
+	c.ticker_index = it->second;
```
Comment (ljflores) [inline src/kv/RocksDBStore_stats.cc:408]: The indentation here looks a little off

File: src/kv/RocksDBStore_stats.cc [hunk 2]
```diff
@@ -0,0 +1,810 @@
...
+      } else {
+	dout(10) << "histogram " << c.name << " not found in rocksdb" << dendl;
```
Comment (ljflores) [inline src/kv/RocksDBStore_stats.cc:410]: Same here with the indentation

File: src/kv/RocksDBStore_stats.cc [hunk 3]
```diff
@@ -0,0 +1,810 @@
...
+    for (auto &c : tickers) {
+      if (mode == all || (mode_bits & c.mode) != 0) {
+	if (db->dbstats) {
+	  uint64_t val = db->dbstats->getTickerCount(c.ticker_index);
+	  f->dump_int(c.name, val);
+	} else {
+	  f->dump_int(c.name, -1);
+	}
+      }
+    }
```
Comment (ljflores) [inline src/kv/RocksDBStore_stats.cc:437]: Indentation is off in this section

File: src/kv/RocksDBStore_stats.cc [hunk 4]
```diff
@@ -0,0 +1,810 @@
...
+	  f->dump_float("p95", -1);
+	  f->dump_float("p99", -1);
+	  f->dump_float("avg", -1);
+	  f->dump_float("std", -1);
+	  f->dump_float("min", -1);
+	  f->dump_float("max", -1);
+	  f->dump_float("count", -1);
+	  f->dump_float("sum", -1);
+	}
+	f->close_section();
```
Comment (ljflores) [inline src/kv/RocksDBStore_stats.cc:470]: Indentation

File: src/kv/RocksDBStore_stats.cc [hunk 5]
```diff
@@ -0,0 +1,810 @@
...
+    double s = 0;
+    double d = 0;
+    for (auto& t : tabs) {
+      // avg = cnt / denominator
+      if (t->tab[avg_index] > 0) {
+	// reconstruct original denominator
+	double denom = t->tab[cnt_index] / t->tab[avg_index];
+	d += denom;
+	s += t->tab[cnt_index];
+	elems++;
```
Comment (ljflores) [inline src/kv/RocksDBStore_stats.cc:524]: Indentation

File: src/kv/RocksDBStore_stats.cc [hunk 6]
```diff
@@ -0,0 +1,810 @@
...
+  void cfstats_sum_everything(ceph::Formatter *f, db_stats_mode mode) {
+    std::vector<cfstats_t> cf_stats_batch;
+    cfstats_t cf_stats;
+    capture_cfstats(db->default_cf, cf_stats);
+    cf_stats_batch.push_back(cf_stats);
+    for (auto& cfh : db->cf_handles) {
+      for (size_t s = 0; s < cfh.second.handles.size(); s++) {
+	cfstats_t cf_stats;
+	capture_cfstats(cfh.second.handles[s], cf_stats);
+	cf_stats_batch.push_back(cf_stats);
```
Comment (ljflores) [inline src/kv/RocksDBStore_stats.cc:694]: Indentation

File: src/kv/RocksDBStore_stats.cc [hunk 7]
```diff
@@ -0,0 +1,810 @@
...
+    capture_cfstats(db->default_cf, cf_stats);
+    dump_cfstats(f, cf_stats, mode);
+    f->close_section();
+
+    for (auto& cfh : db->cf_handles) {
+      cfstats_t cf_stats;
+      std::vector<cfstats_t> cf_stats_batch;
+      for (size_t s = 0; s < cfh.second.handles.size(); s++) {
+	capture_cfstats(cfh.second.handles[s], cf_stats);
+	cf_stats_batch.push_back(cf_stats);
```
Comment (ljflores) [inline src/kv/RocksDBStore_stats.cc:715]: Indentation

File: src/kv/RocksDBStore_stats.cc [hunk 8]
```diff
@@ -0,0 +1,810 @@
...
+      for (size_t s = 0; s < cfh.second.handles.size(); s++) {
+        cfstats_t cf_stats;
+	capture_cfstats(cfh.second.handles[s], cf_stats);
+	if (cfh.second.handles.size() == 1) {
+	  f->open_object_section(cfh.first);
+	} else {
+	  f->open_object_section(cfh.first + "-" + std::to_string(s));
+	}
+	dump_cfstats(f, cf_stats, mode);
+	f->close_section();
```
Comment (ljflores) [inline src/kv/RocksDBStore_stats.cc:741]: Indentation

File: src/kv/RocksDBStore_stats.cc [hunk 9]
```diff
@@ -0,0 +1,810 @@
...
+  }
+  static int create(RocksDBStore* db) {
+    int r = -1;
+    AdminSocket *admin_socket = db->cct->get_admin_socket();
+    if (admin_socket && db->hook == nullptr) {
+      SocketHook* hook = new SocketHook(db);
+      r = admin_socket->register_command(
+	"dump_rocksdb_stats name=level,type=CephString",       
+	hook,
+	"dump_rocksdb_stats <'telemetry'/'objectstore'/'debug'/'all'> ; dump RocksDB stats");
```
Comment (ljflores) [inline src/kv/RocksDBStore_stats.cc:767]: Indentation

File: src/kv/RocksDBStore_stats.cc [hunk 10]
```diff
@@ -0,0 +1,810 @@
...
+      if (r == 0) {
+	db->hook = hook;
```
Comment (ljflores) [inline src/kv/RocksDBStore_stats.cc:769]: Indentation

File: src/kv/RocksDBStore_stats.cc [hunk 11]
```diff
@@ -0,0 +1,810 @@
...
+    int r = 0;
+    if (command == "dump_rocksdb_stats") {
+      std::string level;
+      db_stats::db_stats_mode mode = db_stats::telemetry;
+      if (ceph::common::cmd_getval(cmdmap, "level", level)) {
+	if (level == "telemetry") mode = db_stats::telemetry;
+	else if (level == "objectstore") mode = db_stats::objectstore;
+	else if (level == "debug") mode = db_stats::debug;
+	else if (level == "all") mode = db_stats::all;
+	else return -EINVAL;
```
Comment (ljflores) [inline src/kv/RocksDBStore_stats.cc:789]: Indentation

**Review by ljflores — COMMENTED** [review]
Hey @aclamk, I marked a couple of areas where the indentation seems off. I believe this happens when a mix of spaces and tabs are used, because I've experienced this before where it shows up fine in vim, but not in GitHub.

Looking through your code, it seems well-structured and thorough. I tested this branch on a test cluster (`MON=1 OSD=1 MDS=0 RGW=0 NFS=0 ../src/vstart.sh  -l -n -o rocksdb_perf=true`) and confirmed that I can use this command on osds and mons, all flavors.

In the last telemetry huddle, @markhpc had a chance to look through what this collects for telemetry, and he confirmed that it's what we're looking for. So I believe -- at least in the telemetry sense -- that all of the metrics we need for a good report on RocksDB are being collected.

I see that this branch needs a rebase, so after it's rebased and we've confirmed all the checks have passed, I believe this is good to go for QA, if @neha-ojha and @jdurgin also agree.

**Comment by ljflores** [comment]
@aclamk looks like this PR needs a rebase. Also, about my comments on spacing, I recently found out that there is a difference in the way indentation looks on GitHub depending on which browser you're viewing from (i.e. the indentation looks fine on Chrome, but weird on Firefox). In any case, I would just make sure that the use of tabs/spacing is consistent.

**Comment by ljflores** [comment]
@aclamk if you have time could you rebase this? I'd like to get another review from someone and then see if this can go into testing.

**Comment by ljflores** [comment]
@aclamk I'm closing this since I have resurrected it in a new PR: #45346

**Pull request closed without merging**